File: backend/app/api/v1/predict.py
# pyrefly: ignore [missing-import]
from bson import ObjectId
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api.deps import (
    Caller,
    ensure_self_or_admin,
    get_caller,
    get_current_uid,
    get_database,
)
from app.models.predict_models import (
    CatalogResponse,
    RecommendRequest,
    RecommendResponse,
)

router = APIRouter()

DEFAULT_GENDER = "Nu"


async def resolve_gender(
    caller: Caller, fallback: str | None, db
) -> tuple[str, list[str]]:
    """Giới tính lấy theo thứ tự ưu tiên:

    1. Hồ sơ người dùng đã đăng ký (khi request có token hợp lệ)
    2. Trường `gender` trong body — chỉ dành cho lúc test không có token
    3. Mặc định, kèm cảnh báo

    Không chặn request khi thiếu — mô hình vẫn chạy được, chỉ kém chính xác đi chút.
    """
    warnings: list[str] = []

    if caller.sent_token:
        if caller.uid:
            try:
                doc = await db["users"].find_one({"_id": caller.uid})
                gender = (doc or {}).get("gender")
                if gender in ("Nam", "Nu"):
                    return gender, warnings
                warnings.append(
                    "Hồ sơ chưa có giới tính — hãy cập nhật trong phần tài khoản "
                    "để gợi ý chính xác hơn."
                )
            except Exception:  # noqa: BLE001 — thiếu hồ sơ không được làm hỏng gợi ý
                warnings.append("Không đọc được hồ sơ người dùng.")
        else:
            warnings.append("Token không hợp lệ nên không đọc được hồ sơ.")

    if fallback in ("Nam", "Nu"):
        return fallback, warnings
    return DEFAULT_GENDER, warnings


@router.post(
    "/recommend",
    response_model=RecommendResponse,
    summary="Gợi ý ngành học (XGBoost, mô hình Hướng 1)",
)
async def recommend_majors(
    body: RecommendRequest,
    caller: Caller = Depends(get_caller),
    db=Depends(get_database),
):
    """Mô hình XGBoost Hướng 1 (`research/`, Giai đoạn 10) — 9 nhóm ngành, 63 đặc trưng.

    Đặt `EDUTALK_PIPELINE=r3` để quay về mô hình `research3/`.

    **Hai chế độ** — khác nhau ở chỗ người dùng có chọn nhóm ngành hay không:

    | Chế độ | Khi nào | Nên hiện | Trên tập test | Trên người thật | Đoán bừa |
    |---|---|---|---|---|---|
    | `explore` | `fieldId` bỏ trống | 5 ngành (`limit=5`) | Top-5 81,8% | ~60% | 12,8% |
    | `guided` | `fieldId` = 0..8 | 2 ngành (`limit=2`) | **Top-2 90,6%** | ~79% | 47,6% |

    Tập test gồm 2.546 dòng, phần lớn là hồ sơ trúng tuyển có phần sở thích do mô hình
    sinh; cột "người thật" đo trên phiếu khảo sát (Hướng 2), là mức nên kỳ vọng với
    người dùng. Số trên tập test đo khi tắt lọc mềm theo tổ hợp.

    Đây là hệ **gợi ý** — **không nên hiện `score`** cho người dùng cuối (xác suất
    thật thường thấp, hiện ra sẽ tưởng hệ thống hỏng).

    Gửi kèm `Authorization: Bearer <token>` để server tự lấy giới tính từ hồ sơ.
    """
    from app.services.major_predictor import get_predictor

    predictor = get_predictor()

    gender, warnings = await resolve_gender(caller, body.gender, db)

    try:
        result = predictor.recommend(
            interests=body.interests,
            subject_group=body.subjectGroup,
            gender=gender,
            goal=body.goal,
            scores=body.scores,
            field_id=body.fieldId,
            limit=body.limit,
            loai_bo_ngoai_to_hop=True,
            soft_filter=body.softFilter,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e

    result["warnings"] = warnings + result["warnings"]
    return result


@router.get(
    "/catalog",
    response_model=CatalogResponse,
    summary="Danh mục 9 nhóm ngành và 39 ngành",
)
def get_catalog():
    """Bảng tra khối ngành / ngành lấy thẳng từ mô hình.

    Frontend nên dựng dropdown từ đây thay vì gõ tay danh sách ngành — gõ tay
    chính là lý do bảng gõ tay cũ (đã xoá) lệch 9/39 mã so với mô hình.
    `id` trả về ở đây dùng được luôn cho `fieldId` khi gọi `/recommend`.
    """
    from app.services.major_predictor import get_predictor

    predictor = get_predictor()

    groups: dict[int, dict] = {
        k: {"id": k, "name": v, "subjectGroups": set(), "majors": []}
        for k, v in predictor.field_name.items()
    }
    for i in range(len(predictor.major_name)):
        to_hop = sorted(predictor.to_hop_xet_tuyen.get(i, []))
        g = groups[predictor.field_of_major[i]]
        g["subjectGroups"].update(to_hop)
        g["majors"].append(
            {
                "code": str(predictor.major_code[i]),
                "name": predictor.major_name[i],
                "subjectGroups": to_hop,
                "cutoffs": predictor.diem_chuan.get(i, {}),
            }
        )
    # subjectGroups của khối = hợp của các ngành trong khối. Frontend dùng cái này
    # để chỉ cho chọn tổ hợp thật sự xét tuyển vào khối người dùng đã chọn.
    for g in groups.values():
        g["subjectGroups"] = sorted(g["subjectGroups"])
    return {
        "fields": list(groups.values()),
        "soGoiY": {"tuVan": predictor.so_goi_y(0), "khamPha": predictor.so_goi_y(None)},
    }


# ĐÃ XOÁ: `POST /api/v1/predict/` (19/09/2026).
#
# Endpoint đó không dùng mô hình học máy nào — chỉ chấm cứng 0.85/0.45 theo việc tổ
# hợp có khớp hay không, và 10 điểm khảo sát người dùng nhập vào không hề được đụng
# tới. Nó đọc bảng 39 ngành gõ tay trong `predict_service.py`, bảng này lệch **9/39
# mã ngành** so với mô hình thật, nên client nào lấy mã từ đó rồi gọi sang
# `/recommend` đều không khớp được.
#
# Đã grep toàn bộ `web/src` và `mobile/lib`: không chỗ nào gọi. Thay thế là
# `POST /api/v1/predict/recommend`.


@router.get(
    "/explain/{prediction_id}",
    summary="Đọc lại giải thích SHAP của một lượt tư vấn đã lưu",
)
async def get_explain(
    prediction_id: str,
    uid: str = Depends(get_current_uid),
    db=Depends(get_database),
):
    """Trả về phần giải thích XAI đã lưu kèm lượt tư vấn.

    Chỉ **chính chủ** hoặc **admin** đọc được: bản ghi tư vấn gắn với hồ sơ cá
    nhân, để lộ là lộ luôn điểm thi và sở thích của người khác.

    Thí sinh nhận bản đã lọc bỏ các đặc trưng gắn cờ `anVoiThiSinh` (hiện là giới
    tính). Admin nhận đầy đủ — trang quản trị cần thấy đúng mô hình đã dựa vào gì.
    """
    from bson.errors import InvalidId

    try:
        doc = await db["prediction_history"].find_one({"_id": ObjectId(prediction_id)})
    except (InvalidId, TypeError) as e:
        raise HTTPException(status_code=400, detail="Mã lượt tư vấn không hợp lệ.") from e
    if not doc:
        raise HTTPException(status_code=404, detail="Không tìm thấy lượt tư vấn.")

    la_chu = doc.get("user_id") == uid
    await ensure_self_or_admin(
        doc.get("user_id"), uid, db, "Không có quyền xem lượt tư vấn này."
    )
    la_admin = not la_chu

    majors = []
    for m in doc.get("majors", []):
        gt = m.get("explain")
        if gt and not la_admin:
            gt = {
                **gt,
                "features": [f for f in gt["features"] if not f.get("anVoiThiSinh")],
            }
        majors.append({"rank": m.get("rank"), "name": m.get("name"),
                       "field": m.get("field"), "explain": gt})

    return {
        "id": str(doc["_id"]),
        "mode": doc.get("mode"),
        "thoiGian": doc.get("createdAt").isoformat() if doc.get("createdAt") else None,
        "dayDu": la_admin,
        "majors": majors,
    }

class DienGiaiRequest(BaseModel):
    """Bảng SHAP đã tính sẵn ở `/recommend`, gửi lại để diễn giải thành lời."""

    nganh: str
    features: list[dict]


@router.post(
    "/explain-text",
    summary="Diễn giải bảng SHAP thành 2–3 câu tiếng Việt",
    dependencies=[Depends(get_current_uid)],
)
async def explain_text(body: DienGiaiRequest):
    """Chuyển bảng số thành lời cho học sinh dễ đọc.

    LLM **chỉ được** diễn đạt lại những con số trong `features` — prompt cấm thêm
    lý do, cấm nói về việc làm / lương / điểm chuẩn. Không bao giờ hỏi LLM kiểu
    "đoán xem vì sao mô hình chọn ngành này": cách đó sinh ra lời giải thích nghe
    thuyết phục nhưng không dính gì tới mô hình, tệ hơn là không giải thích.

    Gọi theo yêu cầu (bấm nút) chứ không tự chạy mỗi lần dự đoán — mỗi lượt là một
    lần gọi Gemini, bật sẵn cho cả 5 ngành thì vừa chậm vừa tốn.
    """
    from app.services.xai_service import XAIService

    try:
        loi = await XAIService().dien_giai(body.nganh, body.features)
    except Exception as e:  # noqa: BLE001
        raise HTTPException(
            status_code=503, detail=f"Không tạo được lời giải thích: {e!s}"
        ) from e
    if not loi:
        raise HTTPException(status_code=503, detail="Mô hình ngôn ngữ không trả lời.")
    return {"text": loi}

File: web/src/services/predict.ts
import api from "@/lib/api";

/**
 * Gọi mô hình XGBoost đang phục vụ ở backend (Hướng 1 — `research/`).
 *
 * Số ngành gợi ý KHÔNG gõ cứng ở web: bỏ trống `limit` thì backend tự dùng đúng điểm
 * vận hành mà mô hình được đánh giá (tư vấn 2, khám phá 5). Muốn hiện con số đó trên
 * giao diện thì đọc `soGoiY` từ `catalog()`.
 *
 * Giới tính KHÔNG gửi từ đây — backend tự lấy từ hồ sơ người dùng đã đăng ký.
 * Chỉ truyền `gender` khi test không có token đăng nhập.
 */

// THỨ TỰ 10 CÂU BẮT BUỘC ĐÚNG NHƯ LÚC HUẤN LUYỆN MÔ HÌNH.
// Đổi thứ tự = mô hình đọc sai câu, vẫn chạy nhưng kết quả sai âm thầm.
export const LIKERT_ORDER = [
  "likert_nang_dong",
  "likert_huong_noi",
  "likert_sang_tao",
  "likert_logic",
  "likert_to_mo",
  "likert_thi_nghiem",
  "likert_moi_truong",
  "likert_dinh_duong",
  "likert_tranh_luan",
  "likert_thiet_ke",
] as const;

export const GOAL_BY_ID: Record<number, string> = {
  1: "Đi làm",
  2: "Nghiên cứu",
  3: "Kinh doanh",
  4: "Chưa xác định",
};

export interface AdmissionInfo {
  cutoffs: Record<string, number>;
  min: number;
  max: number;
  latest: number;
  trend: "tang" | "giam" | "on_dinh" | "khong_du_du_lieu";
  level: "an_toan" | "co_kha_nang" | "rui_ro_cao" | null;
  gap: number | null;
}

/** Một dòng trong bảng giải thích SHAP. */
export interface ExplainFeature {
  ten: string;
  giaTri: string;
  /** Đóng góp SHAP của mục này (tổng các cột mã hoá nó) — dương là đẩy lên, âm là kéo xuống */
  dongGop: number;
  phanTram: number;
  /** "rất mạnh" | "mạnh" | "vừa" | "không đáng kể" — do backend chấm, đừng tự tính lại */
  mucDo: string;
  /** Giữ để tương thích — mô hình hiện tại chỉ một tầng nên bằng dongGop */
  tang2: number;
  /** Giữ để tương thích — mô hình hiện tại chỉ một tầng nên luôn bằng 0 */
  tang1: number;
  /**
   * Không hiển thị cho thí sinh — hiện chỉ có giới tính. Nêu một đặc điểm không
   * thể thay đổi vừa không giúp được gì, vừa củng cố định kiến. Vẫn tham gia dự
   * đoán và vẫn trả về để trang quản trị thấy đầy đủ.
   */
  anVoiThiSinh: boolean;
}

export interface MajorExplain {
  mode: string;
  base: number;
  features: ExplainFeature[];
  /** Số đặc trưng không lọt vào danh sách hiển thị */
  soConLai: number;
  dongGopConLai: number;
  /** Tỷ trọng cụm "yếu tố khác"; cộng với phanTram các dòng hiển thị ra 100% */
  phanTramConLai: number;
  /**
   * Tổng đóng góp cả 63 đặc trưng. Điểm xếp hạng = base + tongDongGop.
   * So sánh giữa các ngành phải dùng con số này — KHÔNG so riêng các thanh
   * hiển thị, vì mỗi ngành có `base` riêng và mỗi ngành hiện một bộ 6 khác nhau.
   */
  tongDongGop: number;
}

export interface MajorSuggestion {
  rank: number;
  code: string;
  name: string;
  field: string;
  score: number;
  /** Tổ hợp ngành này xét tuyển (đề án 2026) */
  subjectGroups?: string[];
  /**
   * Ngành có xét tổ hợp thí sinh đã khai không. `false` = điểm thi tổ hợp đó KHÔNG dùng
   * được cho ngành này (chỉ còn học bạ / ĐGNL) — đừng hiện nhãn "trong tầm với" dựa
   * trên tổng điểm ấy. `null` = backend thiếu bảng tuyển sinh, không biết.
   */
  xetToHop?: boolean | null;
  admission: AdmissionInfo | null;
  explain?: MajorExplain | null;
}

export interface RecommendResponse {
  mode: "explore" | "guided";
  fields: { id: number; name: string; probability: number }[];
  majors: MajorSuggestion[];
  totalScore: number | null;
  warnings: string[];
  /** Chỉ có khi `save: true` — mã lượt tư vấn đã lưu, dùng để gắn phản hồi */
  predictionId?: string | null;
}

export interface RecommendInput {
  likertScores: Record<string, number>;
  block: string;
  /** Điểm 3 môn, khóa là id môn (Toan, Ly, Hoa...) */
  scores: Record<string, string>;
  /** Thứ tự môn của tổ hợp — quyết định thứ tự gửi điểm đi */
  subjectOrder: string[];
  goalId: number;
  facultyId: number | null;
  gender?: string;
  /** Bỏ trống = số gợi ý chuẩn của mô hình (tư vấn 2, khám phá 5) */
  limit?: number;
  /**
   * true → gọi `POST /api/v1/survey/submit`: cùng mô hình, cùng kết quả, nhưng
   * server GHI LẠI vào `prediction_history` và tăng `usageCount`. Cần token.
   * false → `POST /api/v1/predict/recommend`: chạy xong là quên, không cần đăng nhập.
   *
   * Trang /history đọc `prediction_history`, nên nếu luôn gọi /recommend thì
   * lịch sử của người dùng sẽ vĩnh viễn rỗng.
   */
  save?: boolean;
}

export interface CatalogField {
  id: number;
  name: string;
  /** Hợp các tổ hợp mà ngành trong khối này xét tuyển */
  subjectGroups: string[];
  majors: CatalogMajor[];
}

export interface CatalogMajor {
  code: string;
  name: string;
  subjectGroups: string[];
  /** Điểm chuẩn thi THPT theo năm, vd { "2024": 20.0, "2025": 22.0, "2026": 19.0 } */
  cutoffs: Record<string, number>;
}

/** Số ngành hiển thị ở mỗi chế độ — đúng điểm vận hành mà mô hình được đánh giá */
export interface SoGoiY {
  tuVan: number;
  khamPha: number;
}

export interface CatalogResponse {
  fields: CatalogField[];
  soGoiY?: SoGoiY | null;
}

export const PredictService = {
  recommend: async (input: RecommendInput): Promise<RecommendResponse> => {
    const interests = LIKERT_ORDER.map((k) => {
      const v = input.likertScores[k];
      if (typeof v !== "number" || v < 1 || v > 5) {
        throw new Error(`Thiếu hoặc sai câu sở thích: ${k}`);
      }
      return v;
    });

    // Điểm phải gửi ĐÚNG thứ tự môn của tổ hợp (vd A00 → Toán, Lý, Hoá)
    const raw = input.subjectOrder.map((id) => parseFloat(input.scores[id] ?? ""));
    const scores = raw.every((v) => !isNaN(v)) ? raw : null;

    const body = {
      interests,
      subjectGroup: input.block,
      scores,
      goal: GOAL_BY_ID[input.goalId] ?? "Chưa xác định",
      fieldId: input.facultyId,
      // undefined bị bỏ khỏi JSON → backend dùng số gợi ý chuẩn của chế độ đang chọn
      limit: input.limit,
      gender: input.gender,
    };

    if (input.save) {
      // `/survey/submit` bọc kết quả trong { status, results }
      const { data } = await api.post<{ results: RecommendResponse; predictionId?: string }>(
        "/api/v1/survey/submit",
        body
      );
      return { ...data.results, predictionId: data.predictionId ?? null };
    }

    const { data } = await api.post<RecommendResponse>(
      "/api/v1/predict/recommend",
      body
    );
    return data;
  },

  catalog: async (): Promise<CatalogResponse> => {
    const { data } = await api.get<CatalogResponse>("/api/v1/predict/catalog");
    return data;
  },
};

/**
 * Nhờ LLM diễn giải bảng SHAP thành 2–3 câu tiếng Việt.
 *
 * Gọi theo yêu cầu (người dùng bấm nút) chứ không tự chạy mỗi lần dự đoán —
 * mỗi lượt là một lần gọi Gemini, bật sẵn cho cả 5 ngành thì vừa chậm vừa tốn.
 *
 * Backend chỉ đưa CHÍNH bảng số này vào prompt và cấm LLM thêm lý do ngoài bảng.
 */
export async function giaiThichBangLoi(
  nganh: string,
  features: ExplainFeature[]
): Promise<string> {
  const { data } = await api.post("/api/v1/predict/explain-text", {
    nganh,
    features,
  });
  return data.text as string;
}
